src/Craftwork/IdObfuscator/CharacterSet.php

```diff
@@ -0,0 +1,131 @@
+<?php
+declare(strict_types=1);
+
+namespace Craftwork\IdObfuscator;
+
+use Craftwork\IdObfuscator\Exception\DuplicateCharacterException;
+use Craftwork\IdObfuscator\Exception\InvalidCharacterSetException;
+use Craftwork\IdObfuscator\Exception\InvalidCharacterSetLengthException;
+
+final class CharacterSet
+{
+    const CHARSET_DEFAULT = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-_';
+
+    /**
+     * @var string
+     */
+    private $characters;
+
+    /**
+     * @param string $characters
+     */
+    private function __construct(string $characters)
+    {
+        $this->assertAllAsciiCharacters($characters);
+        $this->assertCharactersLength($characters);
+        $this->assertNoDuplicateCharacters($characters);
+
+        $this->characters = $characters;
+    }
+
+    /**
+     * @param string $characters
+     *
+     * @return CharacterSet
+     *
+     * @throws InvalidCharacterSetLengthException
+     * @throws DuplicateCharacterException
+     * @throws InvalidCharacterSetException
+     */
+    public static function ofCustomCharacters(string $characters): CharacterSet
+    {
+        return new self($characters);
+    }
+
+    /**
+     * @return CharacterSet
+     */
+    public static function ofDefaultCharacters(): CharacterSet
+    {
+        return new self(self::CHARSET_DEFAULT);
+    }
+
+    /**
+     * @return string
+     */
+    public function getCharacters(): string
+    {
+        return $this->characters;
+    }
+
+    /**
+     * @param string $salt
+     * @return CharacterSet
+     */
+    public function shuffleCharacterSetWithSalt(string $salt): CharacterSet
+    {
+        $saltLength = strlen($salt);
+
+        if ($saltLength === 0) {
+            return $this;
+        }
+
+        $characterSet = $this->characters;
+
+        for ($i = $this->length() - 1, $v = 0, $p = 0; $i > 0; $i--, $v++) {
+            $v %= $saltLength;
+            $int = ord($salt[$v]);
+            $p += $int;
+            $j = ($int + $v + $p) % $i;
+            list($characterSet[$i], $characterSet[$j]) = [$characterSet[$j], $characterSet[$i]];
+        }
+
+        return new self($characterSet);
+    }
+
+    /**
+     * @return int
+     */
+    public function length(): int
+    {
+        return strlen($this->characters);
+    }
+
+    /**
+     * @param string $characters
+     *
+     * @throws InvalidCharacterSetLengthException
+     */
+    private function assertCharactersLength(string $characters)
+    {
+        if ('' === $characters || 2 > strlen($characters)) {
```

Comment:
Wouldn't the 2nd part of the conditional satisfy the 1st? It might be a micro-optimisation to check for the empty string in this way but IMO I'd assign more value to having a simpler if statement